src/partisan_hyparview_peer_service_manager.erl

```diff
@@ -491,37 +509,57 @@ code_change(_OldVsn, State, _Extra) ->
 %%%===================================================================
 
 %% @private
-handle_message({disconnect, Peer}, #state{myself=Myself,
-                                          active=Active0}=State0) ->
-    lager:info("Node ~p received disconnect from ~p", [Myself, Peer]),
+handle_message({disconnect, Peer, DisconnectId},
+               #state{myself=Myself,
+                      active=Active0,
+                      ack_message_map=AckMessageMap0}=State0) ->
+    lager:info("Node ~p received disconnect from ~p with ~p",
+               [Myself, Peer, DisconnectId]),
 
-    %% Remove from active
-    Active = sets:del_element(Peer, Active0),
-    lager:info("Node ~p active view: ~p", [Myself, sets:to_list(Active)]),
+    case is_valid_disconnect(Peer, DisconnectId, AckMessageMap0) of
+        false ->
+            %% Ignore the older disconnect message.
+            {reply, ok, State0};
+        true ->
+            %% Remove from active
+            Active = sets:del_element(Peer, Active0),
+            lager:info("Node ~p active view: ~p", [Myself, sets:to_list(Active)]),
+
+            %% Add to passive view.
+            State = add_to_passive_view(Peer,
+                                        State0#state{active=Active}),
 
-    %% Add to passive view.
-    State = add_to_passive_view(Peer,
-                                State0#state{active=Active}),
+            %% Update the AckMessageMap.
+            AckMessageMap = dict:store(Peer, DisconnectId, AckMessageMap0),
 
-    {reply, ok, State#state{active=Active}};
+            {reply, ok, State#state{ack_message_map=AckMessageMap}}
+    end;
 
-handle_message({neighbor, Peer, Tag, _Sender}, State0) ->
+handle_message({neighbor, Peer, Tag, DisconnectId, _Sender},
+               #state{sent_message_map=SentMessageMap}=State0) ->
     % lager:info("Neighboring with ~p from random walk.", [Peer]),
 
-    %% Add node into the active view.
-    State = add_to_active_view(Peer, Tag, State0),
+    State = case is_addable(DisconnectId, Peer, SentMessageMap) of
+                true ->
+                    %% Add node into the active view.
+                    add_to_active_view(Peer, Tag, State0);
+                false ->
+                    %% @todo Need to send the disconnect message here?
```

Comment:
Similar to the `neighbor_accepted` message, a node added the peer to its active view before sending the `neighbor` message. So, I think we need to let the node know that the `neighbor` message fails to add the node to the peer's active view.